feat:  Ensure the initialization sequence of properties

workgroupengineering opened the pull request:
<!--- See CONTRIBUTING.md for general guidelines on contributions -->

## What does the pull request do?
<!--- Give a bit of background on the PR here, together with links to with related issues etc. -->

Added Xaml Ast Transformer that sort `XamlPropertyAssignmentNode` by `DependsOnAttribute`

## What is the current behavior?
`XamlC ` and `AvaloniaRuntimeXamlLoader.Load`  do not respect `DependsOnAttribute`

## What is the updated/expected behavior with this PR?
`XamlC ` and `AvaloniaRuntimeXamlLoader.Load`  do respect `DependsOnAttribute`


## How was the solution implemented (if it's not obvious)?
<!--- Include any information that might be of use to a reviewer here. -->


## Checklist

- [x] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/avalonia-docs with user documentation

## Breaking changes
<!--- List any breaking changes here. -->

## Obsoletions / Deprecations
<!--- Obsolete and Deprecated attributes on APIs MUST only be included when discussed with Core team. @grokys, @kekekeks & @danwalmsley -->

## Fixed issues
<!--- If the pull request fixes issue(s) list them like this: 
Fixes #123
Fixes #456
-->

 Related to #13248

robloo commented on src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/AvaloniaXamlIlCompiler.cs:
```diff
@@ -92,6 +92,12 @@ void InsertBefore<T>(params IXamlAstTransformer[] t)
+
+            InsertBefore<AvaloniaXamlIlControlTemplatePriorityTransformer>(
+                new DependsOnTransformer()
+                );
```
Can we please follow the convention for this? The entire project and all .NET doesn't do this. This is a new idea to simplify some git change tracking that is completely unnecessary. We should follow what .NET does.

You try to change this in almost every one of your PRs. We need to set a standard and all follow it. In the case of Avalonia and even .NET the standard here seems pretty clear.

workgroupengineering replied:
This is a preliminary PR, I'll reformat it after I'm done.

robloo commented on tests/Avalonia.Markup.Xaml.UnitTests/Xaml/DependsOnAttributeTest.cs:
```diff
@@ -0,0 +1,82 @@
+<Window xmlns='https://github.com/avaloniaui'
+        xmlns:x='http://schemas.microsoft.com/winfx/2006/xaml'
+        xmlns:p='clr-namespace:Avalonia.Markup.Xaml.UnitTests.Xaml'
+        >
```
Again, convention. But I would tolerate more in XAML than C# if it is up to me.

robloo commented on tests/Avalonia.Markup.Xaml.UnitTests/Xaml/DependsOnAttributeTest.cs:
```diff
@@ -0,0 +1,82 @@
+
+    public static readonly DirectProperty<DependsOnControlTest, int> SecondProperty =
+        AvaloniaProperty.RegisterDirect<DependsOnControlTest, int>(nameof(Second)
+            , o => o.Second
+            , (o, v) => o.Second = v);
```
Commas on leading lines looks really strange... convention isn't being respected.

workgroupengineering commented:
Does anyone have any ideas why the test fails on Windows C.I.. It doesn't happen on my local machine.

maxkatz6 commented on tests/Avalonia.Markup.Xaml.UnitTests/Xaml/DependsOnAttributeTest.cs:
```diff
@@ -0,0 +1,94 @@
+    const string Case2_Xaml = @"<Window xmlns='https://github.com/avaloniaui'
+        xmlns:x='http://schemas.microsoft.com/winfx/2006/xaml'
+        xmlns:p='clr-namespace:Avalonia.Markup.Xaml.UnitTests.Xaml;assembly=Avalonia.Markup.Xaml.UnitTests'>
+   <p:DependsOnControlTest x:Name='target' Second='2' First='1'  />
```
DependsOnControlTest is not in Avalonia.Markup.Xaml.UnitTests.Xaml namespace.
It's a nested type, you need to move it out of the nested scope.

workgroupengineering commented:
@maxkatz6 
 i found issue, it is the net 4.7 target.

Avalonia.Markup.Xaml.Loader.dll (netstandard2.0) depends on `System.Memory, Version=4.0.1.1`

Avalonia.Markup.Xaml.UnitTests -> Avalonia.UnitTests -> HarfBuzzSharp 7.3.0 -> System.Memory (>= 4.5.5) using `System.Memory, Version=4.0.1.2`

I tried with binding redirect and but I'm not successful. What do you recommend me to do?

workgroupengineering committed:
test: Ensure the initialization sequence of properties

workgroupengineering committed:
feat: Ensure the initialization sequence of properties

Added Xaml Ast Transformer that sort XamlPropertyAssignmentNode by DependsOnAttribute

workgroupengineering committed:
Merge branch 'master' into fixes/core/DependsOn

workgroupengineering commented:
Do I need to make any other changes?

workgroupengineering committed:
Merge branch 'master' into fixes/core/DependsOn

MrJul commented:
There are two points that I'd like to discuss.


## XAML processing order

The first point is that this PR changes the XAML evaluation order.

AFAIK, and even if it's not specified explicitly, XAML attributes and properties are processed in declaration order (left to right) in all XAML dialects.
Do we really want to change this clear and simple rule?

As an example, given a class `C` with properties `A` and `B`, with `B` having a `[DependsOn("A")]` attribute:
```xaml
<C>
  <C.B><X /></C>
  <C.A><Y /></C>
</C>
```
My understanding is that with this PR, `Y` will now be instantiated before `X`, which the user might not expect at all.
Is it a bit contrived? Probably, but that's still a behavior that can surprise people as it's not obvious.

As a bonus, adding a `DependsOn` attribute now becomes a source breaking change, silently changing the processing order.


## We shouldn't need it

This is the second point, really :)

Controls, especially built-in ones, shouldn't rely on the order their properties are set. (This is *not* in contradiction with the first point.)

This PR will help with a single case: when all members are set at the same time as local values in XAML. We already have a standard mechanism implemented for that very case: `ISupportInitialize` (`BeginInit`/`EndInit`).

If one member in the dependency chain is set through a style or a behavior, `DependsOn` falls apart immediately.

People using plain C# to define their UI should be able to use `BeginInit`/`EndInit` and be done, instead of having to check every property they're setting and ensuring their order matches `DependsOn`.

"Ok, but there are real bugs with `SelectingItemsControl` and property order even thought it implements and handles `ISupportInitialize`!"

Yes, and these should be fixed in `SelectingItemsControl` directly.
Ideally, we'd have some unit tests setting all selection properties in various orders, and all should pass with the same result.

robloo commented:
@MrJul This idea would fix a long standing problem with WPF. If you set the selected item before the items source on some controls it won't work. So I disagree somewhat that it isn't needed entirely. I've seen recent issues about this too. Haven't looked it over in any significant detail though.
